### ISSUE TITLE
Provide an API to configure the TempDirectory used by ASP.NET Core

### ISSUE THREAD

**Issue by pranavkm** [aspnet/AspNetCore#9466]
ASP.NET Core currently uses an environment variable to drive the temp directory, falling back `Path.GetTempPath()`. See https://github.com/aspnet/AspNetCore/blob/56e2080ef7741dea9a5c24a8e0639b0921be2d22/src/Http/Http/src/Internal/BufferingHelper.cs#L18-L39. 

This path is currently in use by `FileBufferingReadStream` and as of https://github.com/aspnet/AspNetCore/pull/9015 change, by `FileBufferingWriteStream`. 

This is one of the few parts of the stack that is only configurable via environment variable. We should instead follow the regular pattern of having an option type (that can be configured via configuration) and use those in when constructing these streams and everywhere else ASP.NET Core may need to write to the temp directory.

**Comment by pranavkm** [comment]
See https://github.com/aspnet/AspNetCore/pull/9015#discussion_r271938569 for disussions. cc @Tratcher

**Comment by davidfowl** [comment]
Tricky as we don’t have a good way to flow this to the places that need this.

**Comment by Tratcher** [comment]
Yes we do. https://github.com/aspnet/AspNetCore/pull/9015#discussion_r275858444

**Comment by davidfowl** [comment]
That doesn't look like a solution. Doesn't that say pass an argument? Isn't this about defaults? What am I missing?

**Comment by Tratcher** [comment]
The options would be consumed by the thing creating the buffer streams such as EnableRewind or MVC's output formatters.
https://github.com/aspnet/AspNetCore/blob/425c196cba530b161b120a57af8f1dd513b96f67/src/Http/Http/src/Internal/BufferingHelper.cs#L41-L51
There was also a proposal to bundle that configuration and stream creation into a service/factory so the individual call sites didn't need to deal with it.

**Comment by anurse** [comment]
This seems reasonable but we don't have resources to address this in 3.0. The basic idea would be to resolve `IOptions<HttpBufferingOptions>` (where `HttpBufferingOptions` is a new type) in the `EnableRewind` method and use it to read this path instead of a global static.

This would be breaking since we'd be deleting (or even just not using) the existing static.

We could take a PR. It's not super complicated. But it will have to land in a major version because it's breaking.

**Comment by Tratcher** [comment]
Note we'd want to update MVC's XmlDataContractSerializerOutputFormatter, XmlSerializerOutputFormatter, and (Newtonsoft) JsonResultExecutor to read these options.

**Comment by cjaliaga** [comment]
In case of resolving `IOptions<HttpBufferingOptions>` in the `EnableRewing` method we could get the it from the request itself `request.HttpContext.RequestServices.GetRequiredService<IOptions<HttpBufferingOptions>>().Value` but, what about the `EnableRewing` for `MultipartSection`?
https://github.com/aspnet/AspNetCore/blob/425c196cba530b161b120a57af8f1dd513b96f67/src/Http/Http/src/Internal/BufferingHelper.cs#L58 

For the streams (`FileBufferingReadStream` and `FileBufferingWriteStream`) should we keep the `Func<string> tempFileDirectoryAccessor` and pass one that reads from the directory specified in the options on the constructors such `EnableRewing, MVC's XmlDataContractSerializerOutputFormatter, XmlSerializerOutputFormatter, (Newtonsoft) JsonResultExecutor...`? 

I'm commenting this because some ctors are now defaulting to the current function with the Environment variable:

```csharp
_tempFileDirectoryAccessor = tempFileDirectoryAccessor ?? AspNetCoreTempDirectory.TempDirectoryFactory;
```
So what is the expected default for the temp directory used in streams when there's no func as param? I can think about:
- Setting one that will use `Path.GetTempPath()`
- Removing the possibility of not passing the func, ensuring that we always pass the func on stream's creation, coming from `IOptions `with `Path.GetTempPath()` as directory if user doesn't explicitly configure it
-  Having a way from the streams to resolve `IOptions`, or a way to get a func from an static that does it. For me it doesn't sound right and not sure about its feasibility.

**Comment by Tratcher** [comment]
That MultipartSection EnableRewind extension is one I hadn't noticed, I wonder where it's used. It already takes `Action<IDisposable> registerForDispose` because it doesn't have direct access to HttpContext, so adding an overload that took the same `Func<string>` as FileBufferingReadStream would work. And yes, the `Func<string>` factory on those types is fine, the caller can get the options for them.

**Comment by Tratcher** [comment]
> So what is the expected default for the temp directory used in streams when there's no func as param?

Those code paths will continue to use the old, static behavior. We can slowly obsolete and remove them.

**Comment by cjaliaga** [comment]
> That MultipartSection EnableRewind extension is one I hadn't noticed, I wonder where it's used. It already takes `Action<IDisposable> registerForDispose` because it doesn't have direct access to HttpContext, so adding an overload that took the same `Func<string>` as FileBufferingReadStream would work. 

I've seen that it's used in the `FormFeature` where we have access to the HttpContext via request, so it should be fine. 

https://github.com/aspnet/AspNetCore/blob/4c79e7fdc017be6ec578f2098168d2646d1cd48d/src/Http/Http/src/Features/FormFeature.cs#L186-L188

>And yes, the `Func<string>` factory on those types is fine, the caller can get the options for them.

I'll work on this and open a PR. Will follow-up if I have any question.

**Comment by cjaliaga** [comment]
Hey @Tratcher, I could use your help here :)

Since the code of the `Func<string>` factory would be the same one used in the callers, I was planning on adding this method to a place that can be shared easily. I thought about adding it to the current `AspNetCoreTempDirectory` class in `Microsoft.AspNetCore.WebUtilities` which seems OK for the `EnableRewind` in `Microsoft.AspNetcore.Http` with the file linked.

```csharp
//Getting options in EnableRewind
var bufferingOptions = request.HttpContext.RequestServices.GetRequiredService<IOptions<HttpBufferingOptions>>().Value;
var fileStream = new FileBufferingReadStream(body, bufferThreshold, bufferLimit, AspNetCoreTempDirectory.TempDirectoryPathFactory(bufferingOptions.TempFileDirectory));
```

However I'm not sure, since currently it's internal and in the Http group of projects, if I should refer to this class from the MVC formatters same way `Microsoft.AspNetcore.Http` is doing, add that factory in another place, just repeat the code in all the callers, or any other option that you may think about.

```csharp
//Function defined in AspNetCoreTempDirectory
public static Func<string> TempDirectoryPathFactory(string path)
{
    return () =>
    {
        // Look for folders in the following order.
        var temp = path ?? Path.GetTempPath();

        if (!Directory.Exists(temp))
        {
            throw new DirectoryNotFoundException(temp);
        }

        return temp;
    };
}
```

Thanks!

**Comment by Tratcher** [comment]
We're slowly trying to replace any APIs in the Internal namespace with fully internal or fully public types. It it's needed by MVC and others then it needs to be in a public namespace.
